x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java

```diff
@@ -0,0 +1,298 @@
+        this.nowSupplier = nowSupplier;
+    }
+
+    public void runPolicy(final String policyId, final ActionListener<EnrichPolicyResult> listener) {
+        // Look up policy in policy store and execute it
+        runPolicy(policyId, enrichStore.getPolicy(policyId), listener);
+    }
+
+    public void runPolicy(final String policyName, final EnrichPolicy policy,
+                          final ActionListener<EnrichPolicyResult> listener) {
+        // TODO: Index Pattern expansion and Alias support
+        // Collect the source index information
+        logger.info("Policy [{}]: Running enrich policy", policyName);
+        final String sourceIndexPattern = policy.getIndexPattern();
+        logger.debug("Policy [{}]: Checking source index [{}]", policyName, sourceIndexPattern);
+        GetIndexRequest getIndexRequest = new GetIndexRequest().indices(sourceIndexPattern);
+        client.admin().indices().getIndex(getIndexRequest, new ActionListener<GetIndexResponse>() {
+            @Override
+            public void onResponse(GetIndexResponse getIndexResponse) {
+                validateMappings(getIndexResponse, policyName, policy, listener);
+                prepareAndCreateEnrichIndex(policyName, policy, listener);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                listener.onFailure(e);
+            }
+        });
+    }
+
+    private Map<String, Object> getMappings(final GetIndexResponse getIndexResponse, final String sourceIndexName) {
+        ImmutableOpenMap<String, ImmutableOpenMap<String, MappingMetaData>> mappings = getIndexResponse.mappings();
+        ImmutableOpenMap<String, MappingMetaData> indexMapping = mappings.get(sourceIndexName);
+        assert indexMapping.keys().size() == 0 : "Expecting only one type per index";
+        MappingMetaData typeMapping = indexMapping.iterator().next().value;
+        return typeMapping.sourceAsMap();
+    }
+
+    private void validateMappings(final GetIndexResponse getIndexResponse, final String policyName, final EnrichPolicy policy,
+                                  final ActionListener<?> listener) {
+        String[] sourceIndices = getIndexResponse.getIndices();
+        logger.debug("Policy [{}]: Validating [{}] source mappings", policyName, sourceIndices);
+        for (String sourceIndex : sourceIndices) {
+            Map<String, Object> mapping = getMappings(getIndexResponse, sourceIndex);
+            Map<?, ?> properties = ((Map<?, ?>) mapping.get("properties"));
+            if (properties == null) {
+                listener.onFailure(
+                    new ElasticsearchException(
+                        "Enrich policy execution for [{}] failed. Could not read mapping for source [{}] included by pattern [{}]",
+                        policyName, sourceIndex, policy.getIndexPattern()));
+            }
+            if (properties.containsKey(policy.getEnrichKey()) == false) {
```

Comment:
we should probably allow `_id` as the enrich key

Reply:
Agreed, though I think it's probably a good idea to put that into a separate PR. I don't think we should bog this one down with too many initial features.